docs/qos_enforcement.md

```diff
@@ -0,0 +1,292 @@
+  max_req_bw_ul: uint32
+  # Guaranteed bit rate
+  # Not enforced by Magma yet
+  gbr:
+    gbr_ul: uint32
+    gbr_dl: uint32
+  # Allocation and retention priority
+  # Not enforced by Magma yet
+  arp:
+    arp_priority_level: uint32
+    arp_preemption_capability: bool
+    arp_preemption_vulnerability: bool
+ # Aggregated Maximum bit rate
+ # for non-GBR bearers, this limits the aggregate bit rate
+ # Enforce this on pipelined
+ apn_ambr:
+   apn_ambr_dl: uint32
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  default_qos_profile: string
+  supported_qos_profiles: array[string]
+  # resource_label should match the name of a specified ApnResource
+  resource_label: string
+```
+
+Multiple QoS objects will be allowed to be associated to APN objects.
+
+One can be specified as the APN-level default policy. This is checked by 
+verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the 
+default QoS object.
+
+Additional QoS objects can be associated to the APN for additional resources 
+beyond the default policy, both physical and logical. 
+For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_resource:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN resource
+  ipv4_block: string 
+  # Specify either VLAN or the interface name for SGi
+  # VLAN to use when communicating over SGi interface
+  vlan_id: string
+  # uplink_bridge must be configured with the uplink interface for this value 
+  # to be valid.
+  uplink_interfaces: string
+  # Prefix for ipv6 allocation
+  ipv6_prefix: string
+```
+
+```
+lte_network:
+  id: string
+  name: string
+  description: string
+  ...
+  apn_resources: array[string]
+  ...
+```
+
+```
+feg_lte_network:
+  id: string
+  name: string
+  description: string
+  ...
+  apn_resources: array[string]
+  ...
+```
+
+For both LTE and federated LTE networks, APN resources must be associated.
+
+```
+subscriber:
+  id: string
+  name: string
+  ...
+  active_base_names: array[base_name]
+  active_policies: array[policy_id]
+  active_apns: array[string]
+  ...
+```
+
+One or more APN objects will be allowed to be associated to subscribers. 
+This is similar to before, though will now create restrictions on associating 
+policies to subscribers.
+
+As before, one or more policies will be allowed to be associated to 
+subscribers. A new limitation will be created where each policy that is 
+associated to a subscriber must share at least one of its QosProfile’s APNs 
+with the subscriber. This must hold true for each policy to subscriber 
+association, and CRUD operations which could break this rule should return 
+an API error and fail to complete.
+
+```
+policy_rule:
+  id: string
+  flow_list: array[flow_description]
+  priority: uint32
+  rating_group: uint32
+  monitoring_key: string
+  ...
+  qos: array[qos_profile]
```

Comment:
There should be only one qos_profile for a given policy rule. Otherwise it won't be clear which QoS rule applies to which flow description.